compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -5587,12 +5593,32 @@ private STNode parseStatementStartsWithIdentifier(SyntaxKind nextTokenKind, STNo
             case SEMICOLON_TOKEN:
                 // Here we directly start parsing as a statement that starts with an expression.
                 return parseStamentStartWithExpr(nextTokenKind, identifier);
+            case QUESTION_MARK_TOKEN:
+                // if the next token is question-mark then it is an optional type descriptor with user defined type
+                switchContext(ParserRuleContext.VAR_DECL_STMT);
+                STNode questionMarkToken = parseQuestionMark();
+                STNode optionalTypeDesc = STNodeFactory.createOptionalTypeDescriptorNode(identifier, questionMarkToken);
+                //check whether if there is more type descriptor elements
+                STNode typeDesc = parseComplexTypeDescriptor(optionalTypeDesc);
+                varName = parseVariableName();
+                finalKeyword = STNodeFactory.createEmptyNode();
+                return parseVarDeclRhs(annots, finalKeyword, typeDesc, varName, false);
             default:
                 // If its a binary operator then this can be a compound assignment statement
                 if (isCompoundBinaryOperator(nextTokenKind)) {
                     return parseCompoundAssignmentStmtRhs(identifier);
                 }
-
+                // If its not a compound assignment statement then it can be union type desc.
+                if (nextTokenKind == SyntaxKind.PIPE_TOKEN) {
+                    switchContext(ParserRuleContext.VAR_DECL_STMT);
+                    STNode pipeToken = parsePipeToken();
+                    STNode rightTypeDesc = parseTypeDescriptor();
+                    STNode unionTypeDesc = STNodeFactory.createUnionTypeDescriptorNode(identifier, pipeToken,
+                            rightTypeDesc);
```

Comment:
Similarly, can we call `parseUnionTypeDescriptor()` here?

Reply:
@SupunS found a better way. Please check.